pkg/credentialprovider/apis/dockerauth/register.go

```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockerauth
```

Comment:
Have you done more thinking about the package and API type names?  Just want to summarize our previous conversation here: basically the original `AuthConfig` struct for holding these credentials was copied from the docker code base, but the reason I had used the name `RegistryCredential` in the KEP and first PR for this is because its current use containing the credentials used for authenticating to registries like ECR, GCR etc doesn't have much to do with docker anymore.

Reply:
I've thought about this quite a bit and I still think `DockerAuth` is the correct term here because the API and interface as you mentioned originated from docker and the keyring is specific to the API we copied from docker https://github.com/moby/moby/blob/master/api/types/auth.go. I know @cheftako had recommneded something like `ContainerRegistryAuth` or `ContainerAuth` though.

Reply:
I know I'm coming in late here, but has there been any thought to leveraging the [Docker credential helper interface](https://github.com/docker/docker-credential-helpers)?  GCR has an [implementation](https://github.com/GoogleCloudPlatform/docker-credential-gcr) and I maintain the one for [ECR](https://github.com/awslabs/amazon-ecr-credential-helper).

Reply:
> I know I'm coming in late here, but has there been any thought to leveraging the Docker credential helper interface? GCR has an implementation and I maintain the one for ECR.

There hasn't, first time I'm seeing this project actually. Is that project designed for managing docker credentials on development machines or for clusters or both?

Also, I think the fact that those projects are named "docker-credentials-*" is justification for naming the auth plugin after docker's credentials apis as well.

Reply:
> Is that project designed for managing docker credentials on development machines or for clusters or both?

I can only speak for the ECR implementation, but that works anywhere that the Docker CLI is invoked.  (Docker only implements the credential helper code in the CLI itself, not in the daemon.)  Amazon ECS has its own code to call the ECR GetAuthorizationToken API directly.

> Also, I think the fact that those projects are named "docker-credentials-*" is justification for naming the auth plugin after docker's credentials apis as well.

The binaries are named that way because the Docker CLI prepends the string "docker-credentials-" to the name of the credential helper configured in `~/.docker/config.json`.  It's an enforced naming requirement by Docker.

Reply:
> The binaries are named that way because the Docker CLI prepends the string "docker-credentials-" to the name of the credential helper configured in ~/.docker/config.json. It's an enforced naming requirement by Docker.

Ah, good to know, thanks.

> I can only speak for the ECR implementation, but that works anywhere that the Docker CLI is invoked. (Docker only implements the credential helper code in the CLI itself, not in the daemon.) Amazon ECS has its own code to call the ECR GetAuthorizationToken API directly.

I think we have to handle the case where the container runtime may not be based on docker (e.g. containerd), but kubelet will pull from a remote registry that supports docker-like authentication.

Reply:
> I think we have to handle the case where the container runtime may not be based on docker (e.g. containerd), but kubelet will pull from a remote registry that supports docker-like authentication. Kubelet doesn't invoke docker's CLI but it does use docker's AuthConfig API internally.

Yep.  Kubelet could implement the same protocol that the Docker CLI uses and then leverage the existing credential helpers that are already out there (including the ones that Docker Desktop vends like the "osxkeychain" helper on Mac OS).

Reply:
Kubelet already uses some docker-specific APIs (like https://github.com/moby/moby/blob/master/api/types/auth.go) but nothing more than that. Any integration that requires invoking docker CLI or expects an environment where docker was installed wouldn't work here. I think a possible implementation of the auth plugin could be based on docker-credential-helpers though :thinking:

Reply:
Yeah, I'm not suggesting that the Kubelet depend on the Docker CLI or even import any of its libraries.  The credential helper protocol is a simple JSON-based interface that uses STDIO to communicate with the helper program and has conventionalized arguments.

Something like this:

```
$ docker-credential-my-helper get <<< "myregistry.example.com"
{"ServerURL":"myregistry.example.com","Username":"foo","Secret":"bar"}
```

There's a brief description of the protocol [here](https://github.com/docker/docker-credential-helpers/blob/master/README.md#development).

Reply:
> Yeah, I'm not suggesting that the Kubelet depend on the Docker CLI or even import any of its libraries. The credential helper protocol is a simple JSON-based interface that uses STDIO to communicate with the helper program and has conventionalized arguments.

Ah, I see what you mean now. So you are suggesting we use the same json-based interface for stdin/stdout communication?

Reply:
I think it's worth evaluating, at least.  There's already a (small) ecosystem of implementations of the credential helper protocol, so if the Kubelet implemented the same interface it'd be able to leverage those existing implementations.  I don't know enough about the Kubelet or Kubernetes to know whether or not it's the right path, but I hadn't seen it mentioned and thought it was worthwhile to ask. :smile:

Reply:
> I don't know enough about the Kubelet or Kubernetes to know whether or not it's the right path, but I hadn't seen it mentioned and thought it was worthwhile to ask

Absolutely, appreciate the mention since I've never seen this. I think my biggest concern would be not including the TypeMeta in the stdio communication. We're gonna be relying on Kubernetes API versioning to evolve the plugin APIs and I'm concerned using the docker-credential-helper would make that difficult. Outside of TypeMeta though, the data being passed into stdin/stdout is identical to what we proposed here, just with different names. Need to steep on this a bit more.

Reply:
Regarding naming, maybe `CRIAuthPlugin` is better?

Reply:
> Regarding naming, maybe CRIAuthPlugin is better?

I thing the flag/type names should not be tied to docker if they can be used against any image registry / container runtime. I haven't through that much about potential names (ImageRegistryAuth, CRIAuth, ...)

Reply:
I like `CRIAuthPlugin` or `RegistryAuthPlugin` (or Credential instead of Auth), albeit Registry slightly better because the authentication is used against the registry and the CRI is more of a middleman.

Reply:
Renamed everything to `CRIAuth*` for now which I think makes sense since the AuthConfig from the keyring lookup is always passed into CRI runtime API:

https://github.com/kubernetes/kubernetes/blob/f4112710f5bbfbd9cc7a68ede0df519af30d5e40/pkg/kubelet/kuberuntime/kuberuntime_image.go#L45-L76